### PR TITLE
Incremental config updates

### PR DESCRIPTION
**Pull request by sukhwinder33445**
requires https://github.com/Icinga/icinga-notifications/pull/191

### CODE REVIEW

**Comment by sukhwinder33445** [comment]
Note to myself:
No changes have been made to the following forms as they will change completely based on the existing PRs:

- SaveEventRuleForm
- EntryForm
- EscalationRecipientForm
- ScheduleForm

File: application/forms/ChannelForm.php [hunk 1]
```diff
@@ -152,50 +153,77 @@ public function hasBeenSubmitted()
...
 
-                return $configItem !== null;
-            },
-            ARRAY_FILTER_USE_BOTH
-        );
+        $channel['config'] = json_encode($this->filterConfig($channel['config']));
+        $storedValues['config'] = json_encode($this->filterConfig($storedValues['config']));
+
+        if(! empty(array_diff_assoc($channel, $storedValues))) {
+            $channel['changed_at'] = time();
```
Comment (oxzi) [inline application/forms/ChannelForm.php:199]: Both the [schema](https://github.com/Icinga/icinga-notifications/blob/24b62333d98d62042be598dc3806e323599527b4/schema/pgsql/upgrades/031.sql#L1-L3) and the [Go code](https://github.com/Icinga/icinga-notifications/blob/24b62333d98d62042be598dc3806e323599527b4/internal/channel/channel.go#L23-L24) are using millisecond precision. Writing seconds in this column results in way smaller values by a factor of 1k, resulting in a non-update from the daemon. Please use millisecond precision for all tables throughout your code.

File: library/Notifications/Model/RuleEscalationRecipient.php [hunk 2]
```diff
@@ -4,49 +4,83 @@
...
 use ipl\Orm\Relations;
+use ipl\Stdlib\Filter;
+
+/**
+ * @property int $id
+ * @property int $rule_escalation_id
+ * @property int $contact_id
+ * @property int $contactgroup_id
+ * @property int $schedule_id
+ * @property int $channel_id
```
Comment (nilmerg) [inline library/Notifications/Model/RuleEscalationRecipient.php:22]: these are all nullable

File: library/Notifications/Model/RuleEscalationRecipient.php [hunk 3]
```diff
@@ -4,49 +4,83 @@
...
+ * @property string $tag
+ * @property string $value
```
Comment (nilmerg) [inline library/Notifications/Model/RuleEscalationRecipient.php:24]: what's this??

File: application/forms/ChannelForm.php [hunk 4]
```diff
@@ -110,14 +114,29 @@ protected function assemble()
...
                 'submit',
                 'delete',
                 [
                     'label'          => $this->translate('Delete'),
                     'class'          => 'btn-remove',
-                    'formnovalidate' => true
+                    'formnovalidate' => true,
+                    'disabled'       => $isInUse,
+                    'title'          => $isInUse
+                        ? $this->translate('Cannot delete channel, contacts or event rules are using it')
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:138]: *Channel is still referenced as a contact's default channel or in an event rule's escalation*

File: application/forms/RotationConfigForm.php [hunk 5]
```diff
@@ -199,97 +202,8 @@ public function __construct(int $scheduleId, Connection $db)
...
-                    Filter::equal('schedule_id', $rotation->schedule_id),
-                    Filter::equal('priority', $rotation->priority),
-                    Filter::greaterThan('first_handoff', $rotation->first_handoff)
-                ))
-                ->orderBy('first_handoff', SORT_ASC)
-                ->first();
-            if ($newerRotation !== null) {
-                $this->nextHandoff = $getHandoff($newerRotation);
-            }
-        }
```
Comment (nilmerg) [inline application/forms/RotationConfigForm.php:279]: This should be kept here. It's not used to populate the form's inputs directly and may be of disadvantage in `fetchDbValues` as if called twice, resets the properties.

File: application/forms/RotationConfigForm.php [hunk 6]
```diff
@@ -468,13 +395,14 @@ public function editRotation(int $rotationId): void
                 }
 
                 if ($lastHandoff === null) {
-                    // If the handoff didn't happen at all, the entry can safely be removed
+                    // If the handoff didn't happen at all, the entry can safely be removed(marked as removed)
                     $this->db->delete('timeperiod_entry', ['id = ?' => $timeperiodEntry->id]);
```
Comment (nilmerg) [inline application/forms/RotationConfigForm.php:399]: What now? Marked or not?

File: application/forms/SaveEventRuleForm.php [hunk 7]
```diff
@@ -383,27 +400,43 @@ function (array $element) use ($recipientId) {
...
+            'is_active' => $values['is_active'] ?? null
+        ]);
+
+        if (array_key_exists('object_filter', $values)) {
+            $data['object_filter'] = $values['object_filter'];
+        }
+
+        $changedAt = time() * 1000;
+        if (! empty($data)) {
+            $db->update('rule', $data, ['id = ?' => $id]);
```
Comment (nilmerg) [inline application/forms/SaveEventRuleForm.php:424]: Doesn't this miss `changedAt`?

**Review by nilmerg — CHANGES_REQUESTED** [review]
forms only

File: application/forms/RotationConfigForm.php [hunk 8]
```diff
@@ -245,12 +231,25 @@ public function loadRotation(int $rotationId): self
...
+            /** @var ?Rotation $rotation */
+            $rotation = Rotation::on($this->db)
+                ->filter(Filter::all(
+                    Filter::equal('id', $this->rotationId),
+                    Filter::equal('deleted', 'n')
+                ))
+                ->first();
+            if ($rotation === null) {
+                throw new HttpNotFoundException($this->translate('Rotation not found'));
+            }
```
Comment (nilmerg) [inline application/forms/RotationConfigForm.php:243]: This shouldn't be required. Just update the code to access what `fetchDbValues()` returns, please.

File: library/Notifications/Model/RuleEscalationRecipient.php (hunk 3 above)
Comment (nilmerg) [reply library/Notifications/Model/RuleEscalationRecipient.php:24]: These are still here!

File: application/controllers/SourcesController.php [hunk 9]
```diff
@@ -93,9 +96,8 @@ public function addAction(): void
     {
         $form = (new SourceForm(Database::get()))
             ->on(SourceForm::ON_SUCCESS, function (SourceForm $form) {
-                /** @var string $sourceName */
-                $sourceName = $form->getValue('name');
-                Notification::success(sprintf(t('Added new source %s has successfully'), $sourceName));
+                $form->addSource();
+                Notification::success(sprintf(t('Add new source %s successfully'), $form->getSourceName()));
```
Comment (nilmerg) [inline application/controllers/SourcesController.php:100]: ~~Add~~**Added**

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — CHANGES_REQUESTED** [review]
Looks okay now. Though, please try to centralize the registration of `deleted=n` in queries.

Take a look at the `Database` trait and it's `ON_ASSEMBLE_SELECT` listener. There you should be able to check a select whether it queries one of the tables with this column and add the filter for it.

You might need to ignore selects which already filter explicitly for `deleted=y`.

File: application/forms/ContactGroupForm.php [hunk 10]
```diff
@@ -235,34 +235,51 @@ public function editGroup(): bool
...
         }
 
         if (! empty($toAdd)) {
+            $contactsMarkedAsDeleted = $this->db->fetchCol(
+                (new Select())
+                    ->from('contactgroup_member')
+                    ->columns(['contact_id'])
+                    ->where([
+                        'contactgroup_id = ?'   => $this->contactgroupId,
+                        'contact_id IN (?)'     => $toAdd
```
Comment (nilmerg) [inline application/forms/ContactGroupForm.php:255]: Didn't this filter for `deleted=y` previously?

File: library/Notifications/Widget/Timeline/Rotation.php [hunk 11]
```diff
@@ -77,6 +77,10 @@ public function fetchTimeperiodEntries(DateTime $after, DateTime $until): Genera
         $entries = $this->model->timeperiod->timeperiod_entry
             ->with(['member.contact', 'member.contactgroup'])
             ->filter(Filter::all(
+                Filter::any(
+                    Filter::equal('member.contact.deleted', 'n'),
+                    Filter::equal('member.contactgroup.deleted', 'n')
+                ),
```
Comment (nilmerg) [inline library/Notifications/Widget/Timeline/Rotation.php:83]: Can be removed as discussed.

**Review by nilmerg — CHANGES_REQUESTED** [review]

File: library/Notifications/Model/Rotation.php [hunk 12]
```diff
@@ -99,4 +104,60 @@ public function retrieve(Model $model): void
             }
         });
     }
+
+    /**
+     * Delete rotation and all related references
+     *
+     * The call must be wrapped in a transaction
```
Comment (nilmerg) [inline library/Notifications/Model/Rotation.php:111]:
```php
$transationStarted = false;
if (! $db->inTransaction()) {
    $transactionStarted = true;
    $db->beginTransaction();
}

...

if ($transactionStarted) {
    $db->commitTransaction();
}
```

File: library/Notifications/Model/Rotation.php [hunk 13]
```diff
@@ -99,4 +104,60 @@ public function retrieve(Model $model): void
...
+     *
+     * @return void
+     */
+    public function delete(): void
+    {
+        if (! $this->timeperiod instanceof Timeperiod) {
+            throw new InvalidArgumentException('Timeperiod must be loaded with the rotation');
+        }
```
Comment (nilmerg) [inline library/Notifications/Model/Rotation.php:119]:
```php
if ($this->timeperiod instanceof Timeperiod) {
    $timeperiod = $this->timeperiod;
} else {
    $timeperiod = $this->timeperiod->first();
}
```

File: library/Notifications/Model/Rotation.php (hunk 12 above)
Comment (nilmerg) [reply library/Notifications/Model/Rotation.php:110]: The comment is now obsolete.

File: library/Notifications/Model/Rotation.php [hunk 14]
```diff
@@ -91,4 +103,71 @@ public function retrieve(Model $model): void
...
+        $transactionStarted = false;
+        if (! $db->inTransaction()) {
+            $transactionStarted = true;
+            $db->beginTransaction();
+        }
+
+        if ($this->timeperiod instanceof Timeperiod) {
+            $timeperiodId = $this->timeperiod->id;
+        } else {
+            $timeperiodId = $this->timeperiod->columns('id')->first();
```
Comment (nilmerg) [inline library/Notifications/Model/Rotation.php:126]: Above you fetch the id, here the entire row.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — COMMENTED** [review]
:hourglass_flowing_sand:

File: application/forms/ChannelForm.php [hunk 15]
```diff
@@ -110,14 +114,36 @@ protected function assemble()
...
+                ->columns('1')
+                ->filter(Filter::equal('default_channel_id', $this->channelId))
+                ->first();
+
+            if ($isInUse === null) {
+                $isInUse = RuleEscalationRecipient::on($this->db)
+                    ->columns('1')
+                    ->filter(Filter::any(
+                        Filter::equal('channel_id', $this->channelId),
+                        Filter::equal('contact.default_channel_id', $this->channelId)
```
Comment (nilmerg) [inline application/forms/ChannelForm.php:127]: The second filter is already covered by the first query above.

**Review by nilmerg — CHANGES_REQUESTED** [review]

**Review by nilmerg — APPROVED** [review]

**Pull request merged**